app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -102,258 +101,296 @@ public DataPullTask(String username, String password,
-                    }
-                } else if (pullResponse.responseCode == 500) {
-                    if (loginNeeded) {
-                        CommCareApplication._().releaseUserResourcesAndServices();
-                    }
-                    Logger.log(AndroidLogger.TYPE_USER, "500 Server Error|" + username);
-                    return new ResultAndError<>(PullTaskResult.SERVER_ERROR, "");
-                }
-            } catch (SocketTimeoutException e) {
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Timed out listening to receive data during sync");
-                responseError = PullTaskResult.CONNECTION_TIMEOUT;
-            } catch (ConnectTimeoutException e) {
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Timed out listening to receive data during sync");
-                responseError = PullTaskResult.CONNECTION_TIMEOUT;
-            } catch (ClientProtocolException e) {
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Couldn't sync due network error|" + e.getMessage());
-            } catch (UnknownHostException e) {
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Couldn't sync due to bad network");
-                responseError = PullTaskResult.UNREACHABLE_HOST;
-            } catch (IOException e) {
-                // TODO Auto-generated catch block
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Couldn't sync due to IO Error|" + e.getMessage());
             }
-            if (loginNeeded) {
-                CommCareApplication._().releaseUserResourcesAndServices();
+        };
+    }
+
+    private byte[] getEncryptionKey() {
+        byte[] key;
+        if (loginNeeded) {
+            initUKRForLogin();
+            if (ukrForLogin == null) {
+                return null;
+            }
+            key = ByteEncrypter.wrapByteArrayWithString(ukrForLogin.getEncryptedKey(), password);
+        } else {
+            //Only purge cases if we already had a logged in user. Otherwise we probably can't read the DB.
+            CaseUtils.purgeCases();
+            key = CommCareApplication._().getSession().getLoggedInUser().getWrappedKey();
+        }
+        this.publishProgress(PROGRESS_CLEANED); // Either way, we don't want to do this step again
+        return key;
+    }
+
+    private void initUKRForLogin() {
+        if (shouldGenerateFirstKey()) {
+            SecretKey newKey = CryptUtil.generateSemiRandomKey();
+            if (newKey == null) {
+                return;
             }
+            String sandboxId = PropertyUtils.genUUID().replace("-", "");
+            ukrForLogin = new UserKeyRecord(username, UserKeyRecord.generatePwdHash(password),
+                    newKey.getEncoded(), new Date(), new Date(Long.MAX_VALUE), sandboxId);

```

Comment:
This line represents an actual change from the old code https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/tasks/DataPullTask.java#L181, where we were wrapping the encrypted key before passing it into the `UserKeyRecord` constructor. I removed that operation here because I'm pretty sure that that was incorrect; the `UserKeyRecord` constructor does not indicate that it expects a wrapped key, and we perform the wrapping function again on line 209 above after we extract it from the ukr. But @ctsims you should double check and confirm.